<div style="background-color:#f9f9f9; padding:10px; border:1px solid #ddd; border-radius:8px;">

<h1 style='text-align:center;color:#2c3e50;font-family:Verdana;'><strong>Big Data Science - Fall 2023</strong></h1>

<h2 style='text-align:center;color:#3498db;font-family:Verdana;'>Assignment 2: Large Language Models on Protein Sequences</h2>

<hr style="border-top: 2px solid #3498db; margin-bottom: 20px; margin-top: 20px;">

<h3 style='color:#2c3e50;font-family:Verdana;'>Milestones:</h3>

<ol type="1" style='font-family:Verdana;'>
    <li><span style='color:#e74c3c;'>Understanding and Implementing Attention Mechanism and BERT Architecture</span></li>
    <li><span style='color:#27ae60;'>Fine-tuning Model on 2 Biological Tasks</span></li>
    <li><span style='color:#3498db;'>Evaluating Models' Performance with or without pretraining</span></li>
</ol>

</div>


<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h3 style='color:#2c3e50;'><strong>Submission Deadline: Friday, December 8 at 8:59 P.M.</strong></h3>

<p style='color:#2c3e50;'><strong>Notes before you embark on this assignment journey:</strong></p>

<ul style='color:#2c3e50;'>
    <li><strong>Avoid using built-in libraries</strong> for the core algorithms, strive to build from scratch to strengthen your understanding. If you are allowed to use built-in libraries, you will be notified in the respective sections.
    </li>
    <li><strong>Collaboration is key</strong> to successful learning, feel free to discuss problems with your peers but ensure your submissions are individual efforts.</li>
    <li><strong>The early bird catches the worm!</strong> Start early to allow ample time for exploring, implementing, and seeking help if needed.</li>
    <li><strong>Ensure your submission</strong> is a well-documented Jupyter Notebook. Attach your *.ipynb file under Assignment 2 on Canvas.</li>
</ul>

</div>

<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h2 style='color:#2c3e50;'><strong>Section 1: Problem Definition</strong></h2>

<hr style="border-top: 2px solid #3498db; margin-bottom: 20px; margin-top: 20px;">

<h3 style='color:#2c3e50;'>Goal</h3>
<p style='color:#2c3e50;'>Dive into the realm of Molecular Biology and explore how a Language Model (LM) like BERT can be harnessed for protein sequences. </p>

<h3 style='color:#2c3e50;'>Introduction to the Problem Domains</h3>
<ul style='color:#2c3e50;'>
    <li><strong>Proteins</strong> are the workhorses of the body, performing a wide range of functions and biological processes necessary for survival and wellbeing. Understanding their functions can aid in drug discovery, disease diagnosis, and much more. 
    Proteins are at least 50, and usually more than 100, amino acids in length and composed of multiple peptide subunits. Each protein consists of a linear sequence of amino acids. The sequence of a protein is usually notated as a string of letters including 22 aminoacids and are varied in length from very short to very long! For instance, in NLP we use words and sentences as sequences, similarly in protein modeling we also work with sequences which looks something like this: 
    <br>​
    V​​L​​S​​P​​A​​D​​K​​T​​N​​VK​​A​​A​​W​​G​​K​​V​​G​​A​​H​A​​G​​E​​Y​​G​​A​​E​​A​​L​​E​R​​M​​F​​L​​S​​F​​P​​T​​T​​KT​​Y​​F​​P​​H​​F​​D​​L​​S​​HG​​S​​A​​Q​​V​​K​​G​​H​​G​​K​K​​V​​A​​D​​A​​L​​T​​N​​A​​VA​​H​​V​​D​​D​​M​​P​​N​​A​​L​S​​A​​L​​S​​D​​L​​H​​A​​H​​K​​L​​R​​V​​D​​P​​V​​N​​F​​K​​L​L​​S​​H​​C​​L​​L​​V​​T​​L​​AA​​H​​L​​P​​A​​E​​F​​T​​P​​A​V​​H​​A​​S​​L​​D​​K​​F​​L​​A​​S​​V​​S​​T​​V​​L​​T​​S​​K​​Y​ 
    </br>
    <img alt="Protein Structure" src="protein.jpeg" width="600" height="400" align="center"/>
    <br>
    This sequence belongs to a structure that looks like this: </br>
    </li>
    <img alt="Protein Structure" src="structure.jpg" width="600" height="400" align="center"/>
    <br>
    You can find more information about proteins <a href="https://www.khanacademy.org/science/biology/macromolecules/proteins-and-amino-acids/v/introduction-to-amino-acids">here</a>.
    </br>
    <li><strong>Signal Peptides (SP)</strong> are short sequences of amino acids, typically between 2 and 50 amino acids in length, within a protein that direct the protein to specific locations within or outside the cell. They act like postal addresses, guiding the cellular delivery machinery to transport the protein to its intended location.Once the protein reaches its destination, the signal peptide is often cleaved off by specific enzymes. In the following image, you can see the difference between a peptide and a protein: 
    <br>
    <img alt="protein vs peptide" src="peptide.jpeg" width="600" height="400" align="center"/>
    <br>
    You can find more information about peptides <a href="https://www.cellgs.com/blog/the-difference-between-peptides-and-proteins.html">here</a>.
    <br>
    We have a dataset for signal peptides detection in "data/signal_peptide" directory. In this dataset, if a protein sequence has a signal peptide, it is labeled as 1, otherwise it is labeled as 0.
    </br>
    </li>
    <li><strong>SCOP</strong>, which stands for Structural Classification of Proteins is a largely manual classification of protein structural domains based on similarities of their structures and amino acid sequences. Predicting the structural class of a protein sequence is another fine-tuning task that we will do in this assignment. For each protein sequence, we have its structural class as its label in "data/scop" directory. Labels are categorical and there are 7 classes in total.
    </li>
</ul>
</div>

<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h3 style='color:#2c3e50;'><strong>Import Libraries and Read Data</strong></h3>


In [1]:
# Please do not change this cell. Run it without changes.
# This cell imports the required libraries and sets some parameters.
# If you want to change the parameters, you can do so here and mention why you changed them with a comment.

import numpy as np
import pandas as pd
import tensorflow as tf
import os
import math
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

max_length = 256 # size of the input sequences - if a sequence is longer than this, it will be truncated. If it's shorter, it will be padded.
batch_size = 32  # number of sequences that will be given to the model at once
epochs = 10     # number of times the model will see the whole training set
lr = 1e-2      # learning rate
vocab_size = 26  # size of the vocabulary - do not change 
embed_size = 128 # size of the embeddings - do not change
n_heads = 4     # number of attention heads - do not change
n_layers = 2   # number of transformer layers - do not change

In [2]:
# please don't change this cell. Run it without changes.
# this cell is for loading the data needed for the assignment. 
# pretrain data is downloaded from UniprotKB-SwissProt (https://www.uniprot.org/downloads) latest release.

pretrain_data = pd.read_csv('data/pretrain.tsv', sep='\t')
# scop data
scop_train_data = pd.read_csv('data/scop/scop.train.csv')
scop_test_data = pd.read_csv('data/scop/scop.test.csv')
# signal peptide data
signalp_train_data = pd.read_csv('data/signal_peptide/signalP_binary.train.csv')
signalp_test_data = pd.read_csv('data/signal_peptide/signalP_binary.test.csv')

In [3]:
# please don't change this cell. Run it without changes.
# this cell is for looking at the pretrain data to get an idea of what it looks like.

pretrain_data.head()

id                                                seq
0  Q65W17  MKPLVIKLGGVLLDTPAAMENLFTALADYQQNFARPLLIVHGGGCL...
1  Q12ZI9  MFTILTGSQFGDEGKGKIVDLLSKDYDLVVRFQGGDNAGHTVVVGD...
2  O08400  MTRISRSAYAEIYGPTVVGGVGDRVRLADTLLLAEVEKDHTIFGEE...
3  O14232  MFGGELDDAFGVFEGKVPKSLKEESKNSQNSQNSQKIKRTLTDKNA...
4  Q2FQ95  MNILIVNRYGDPDVEEFSYELEKLLHHHGHHTSIYKENLLGEAPPL...

In [4]:
# please don't change this cell. Run it without changes.
# this cell is for looking at the scop data to get an idea of what it looks like.

scop_train_data.head()

seq label
0  MSPFTGSAAPTPEWRHLRVEITDGVATVTLARPDKLNALTFEAYAD...     c
1  MVVTKLAPDFKAPAVLGNNEVDEHFELSKNLGKNGVILFFWPKDFT...     c
2  MKVGIDAGGTLIKIVQEQDNQRTFKTELTKNIDQVVEWLNQQQIEK...     c
3  LYKLLILDIDGTLRDEVYGIPESAKHAIRLCQKNHCSVVICTGRSM...     c
4  NTSNITFIGGGNMARNIVVGLIANGYDPNRICVTNRSLDKLDFFKE...     c

In [5]:
# please don't change this cell. Run it without changes.
# this cell is for looking at the signal peptide data to get an idea of what it looks like.

signalp_train_data.head()

label                                                seq
0      0  MLGMIRNSLFGSVETWPWQVLSTGGKEDVSYEERACEGGKFATVEV...
1      1  MQPAKNLLFSSLLFSSLLFSSAARAASEDGGRGPYVQADLAYAAER...
2      0  MDKGEGLRLAATLRQWTRLYGGCHLLLGAVVCSLLAACSSSPPGGV...
3      0  MKFIDEAKIEVAAGKGGNGATSFRREKFVPRGGPDGGDGGKGGSVW...
4      0  MVAGMLMPRDQLRAIYEVLFREGVMVAKKDRRPRSLHPHVPGVTNL...

<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h3 style='color:#2c3e50;'><strong>Section 1. Tokenize Protein Sequences</strong></h3>

<p style='color:#2c3e50;'>Firstly, you need to have your protein sequences in a format that can be fed into the model. Each amino acid can be represented by a unique token (or ID), similar to how each word is represented by a unique token in NLP tasks. A common representation is to use the single-letter codes for amino acids.</p>
<p style='color:#2c3e50;'>We have <b>22 amino acids: "ACDEFGHIKLMNPQRSTUVWXY" </b>. There also can be invalid amino acids in protein sequences that we can encode them by <b>'OTHER' </b>token. We need tokens <b>'START'</b>, <b>'END'</b>, and <b>'PAD'</b> for showing starting of the protein sequence, end of protein sequence and padding the protein sequence to a fixed size. <p><b> Do not use predefined tokenizer in keras or tensorflow. You need to implement your own tokenizer for protein sequences.</b></p>
</p>
<p>
    <b>Example (padding to size 32):</b> <br>
    <b>Label encoding:</b> {A: 0, C: 1, D: 2, E: 3, F: 4, G: 5, H: 6, I: 7, K: 8, L: 9, M: 10, N: 11, P: 12, Q: 13, R: 14, S: 15, T: 16, V: 17, W: 18, X: 19, Y: 20, OTHER: 21, START: 22, END: 23, PAD: 24} <br>
    <b>Input:</b> "ACDEFGHIKLMNPQRSTUVWXY" <br>
    <b>Output:</b> [22, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10,11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 23, 24, 24, 24, 24, 24, 24, 24, 24, 24] <br>
</p>
</div>


In [6]:
def tokenize_seq(seq, max_length=512):
    '''
    Tokenize a sequence into a list of integers.
    assign each amino acid to a unique integer and add <START>, <END>, <PAD>, <OTHER> tokens to the vocabulary.
    <START> token is added to the beginning of each sequence
    <END> token is added to the end of each sequence
    <PAD> token is used to pad sequences to the same length (max_length)
    <OTHER> token is used to represent all amino acids that are not in the vocabulary
    
    Args:
        seq (str): protein sequence
        max_length (int): maximum length of the sequence
    Returns:
        tokenized_seq: list of integers representing the sequence
    '''
    # Define a dictionary for label encoding including 'START', 'END', 'PAD' and 'OTHER'
    label_encoding = {
        'A': 0, 'C': 1, 'D': 2, 'E': 3, 'F': 4, 'G': 5, 'H': 6, 'I': 7, 'K': 8, 'L': 9,
        'M': 10, 'N': 11, 'P': 12, 'Q': 13, 'R': 14, 'S': 15, 'T': 16, 'V': 17, 'W': 18,
        'X': 19, 'Y': 20, 'OTHER': 21, 'START': 22, 'END': 23, 'PAD': 24
    }
    
    # Start with the 'START' token
    tokenized_seq = [label_encoding['START']]
    
    # Tokenize the sequence
    for acid in seq:
        tokenized_seq.append(label_encoding.get(acid, label_encoding['OTHER']))  # Use 'OTHER' if acid not found
    
    # Add the 'END' token
    tokenized_seq.append(label_encoding['END'])
    
    # Pad the sequence to the max_length with 'PAD' tokens
    tokenized_seq.extend([label_encoding['PAD']] * (max_length - len(tokenized_seq)))
    
    # Truncate the sequence if it's longer than max_length
    return tokenized_seq[:max_length]


In [7]:
# please don't change this cell and don't add any print statements to this cell. Run it without changes.
# this cell is tokenizing the pretrain data and converting it to numpy array.

pretrain_seq = [tokenize_seq(seq) for seq in pretrain_data['seq']]
pretrain_seq = np.array(pretrain_seq)
pretrain_seq = pretrain_seq.astype(np.int32)

<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h2 style='color:#2c3e50;'><strong>Section 2: Pretrain a Language Model</strong></h2>

<hr style="border-top: 2px solid #3498db; margin-bottom: 20px; margin-top: 20px;">

<ul style='color:#2c3e50;'>
    <li>Implement the attention mechanism</li>
    <li>Implement BERT architecture</li>
    <li>Train the model by the protein sequences.</li>
</ul>

</div>

<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h3 style='color:#2c3e50;'><strong>Task 2.1: Implementing the Self-Attention Class</strong></h3>

<p style='color:#2c3e50;'>Implement a multihead self-attention mechanism that will be the building block for BERT model.</p>

<p style='color:#2c3e50;'><strong>Multihead self-attention</strong> is a mechanism at the heart of Transformer models, which are widely used for various natural language processing tasks. It allows the model to focus on different positions of the input sequence simultaneously when producing an output. Here's a simplified explanation:
<p style='color:#2c3e50;'>
<strong>Self-Attention:</strong> This component allows each position in a sequence to attend to all positions within the same sequence when computing the representation of itself. It helps the model capture context from the entire sequence.
<p style='color:#2c3e50;'>
<strong>Multihead:</strong> The 'multihead' part means that the self-attention process is duplicated multiple times (each 'head' being one instance). Each head attends to information from different representational spaces at different positions. This means that instead of having a single set of attention weights for each position, you have multiple sets, allowing the model to capture a diverse range of information which enhances its learning capacity for complex patterns in data.
<p style='color:#2c3e50;'>
<strong>Mechanism Overview:</strong> In practice, for each head, the input sequence is linearly transformed to a set of queries, keys, and values using learned weights. Then, the attention scores are computed by a <b>scaled dot-product of queries and keys</b>, which are used to weight the values. The outputs of each head are concatenated and once again linearly transformed to produce the final output. </p>
<img alt="Multihead Self-Attention" src="attention.jpg" width="600" height="400" align="center"/>
<p> You can read more about attention mechanism <a href="https://towardsdatascience.com/transformers-explained-visually-part-3-multi-head-attention-deep-dive-1c1ff1024853">here</a>.</p>
<p>
An attention mask in the context of self-attention mechanisms is used to prevent certain positions from being attended to. For instance, when processing sequences of different lengths, padding tokens are added to achieve uniformity in length. An attention mask can be applied to ensure that these padding tokens do not influence the model's output. It's also used to enforce causality in sequence-to-sequence tasks, ensuring that predictions for a certain position can't depend on future positions. Although it is necessary to use attention masks in some cases, we will not use them in this assignment.
</div>

In [8]:
class MultiSelfAttentionHead(tf.keras.layers.Layer):
    def __init__(self, num_heads=4, key_dim=64, **kwargs):
        super(MultiSelfAttentionHead, self).__init__(**kwargs)
        self.num_heads = num_heads
        self.key_dim = key_dim
        self.scale = key_dim ** -0.5  # Scale factor for the dot product
        self.depth = key_dim // num_heads  # Depth of each head

        assert key_dim % num_heads == 0, "key_dim must be divisible by num_heads"

    def build(self, input_shape):
        # Create the trainable weight matrices for Q, K, V and the final output in this function.
        self.wq = self.add_weight(shape=(input_shape[-1], self.key_dim), initializer='random_normal', trainable=True, name='wq')
        self.wk = self.add_weight(shape=(input_shape[-1], self.key_dim), initializer='random_normal', trainable=True, name='wk')
        self.wv = self.add_weight(shape=(input_shape[-1], self.key_dim), initializer='random_normal', trainable=True, name='wv')
        self.wo = self.add_weight(shape=(self.key_dim, input_shape[-1]), initializer='random_normal', trainable=True, name='wo')
        super(MultiSelfAttentionHead, self).build(input_shape)

    def split_heads(self, x, batch_size):
        # Split the last dimension into (num_heads, depth)
        x = tf.reshape(x, (batch_size, -1, self.num_heads, self.depth))
        # Transpose to shape (batch_size, num_heads, seq_len, depth)
        return tf.transpose(x, perm=[0, 2, 1, 3])
    
    def call(self, inputs):
        batch_size = tf.shape(inputs)[0]
        
        # Linear projections
        query = tf.matmul(inputs, self.wq)  # (batch_size, seq_len, key_dim)
        key = tf.matmul(inputs, self.wk)  # (batch_size, seq_len, key_dim)
        value = tf.matmul(inputs, self.wv)  # (batch_size, seq_len, key_dim)

        # Split heads
        query = self.split_heads(query, batch_size)  # (batch_size, num_heads, seq_len_q, depth)
        key = self.split_heads(key, batch_size)  # (batch_size, num_heads, seq_len_k, depth)
        value = self.split_heads(value, batch_size)  # (batch_size, num_heads, seq_len_v, depth)

        # Scaled dot-product attention
        matmul_qk = tf.matmul(query, key, transpose_b=True)  # (batch_size, num_heads, seq_len_q, seq_len_k)
        scaled_attention_logits = matmul_qk * self.scale
        # Apply softmax to the last axis
        attention_weights = tf.nn.softmax(scaled_attention_logits, axis=-1)  # (batch_size, num_heads, seq_len_q, seq_len_k)

        # Attention context vector (value weighted by the attention scores)
        context = tf.matmul(attention_weights, value)  # (batch_size, num_heads, seq_len_q, depth)
        # 'Concatenate' heads
        context = tf.transpose(context, perm=[0, 2, 1, 3])  # (batch_size, seq_len_q, num_heads, depth)
        context = tf.reshape(context, (batch_size, -1, self.key_dim))  # (batch_size, seq_len_q, key_dim)

        # Final linear projection
        outputs = tf.matmul(context, self.wo)  # (batch_size, seq_len_q, input_shape[-1])
        return outputs

<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h3 style='color:#2c3e50;'><strong>Task 2.2: Implementing the BERT Architecture</strong></h3>

<p style='color:#2c3e50;'>Implement BERT architecture.</p>
<p> You can read more about BERT in the original <a href="https://arxiv.org/pdf/1810.04805.pdf">paper</a>.</p>
<p>
BERT is based on encoder transformer. Encoder consists of multi-head attention and feed-forward neural network. In original Transformers paper <a href="https://arxiv.org/pdf/1706.03762.pdf">Attention Is All You Need</a> and in this assignment, the stack of identical encoder layers is used. 
</p>
<img alt="Encoder transformer" src="encoder.png" width="300" height="400" align="center"/>
<p> In BERT function, you need to implement the following steps:</p>
<p>
<strong>Inputs:</strong> The function takes several parameters like vocabulary size, embedding size, number of layers, attention heads, and maximum sequence length. Additionally, it expects an input sequence and an attention mask when called.
<p>
<strong>Embedding Layer:</strong> The model begins by mapping the input sequence of tokens to vectors using an embedding layer.
<p>
<strong>Encoder Layers:</strong> It then passes the embedded input through a series of identical layers. Each layer consists of:
1. A multi-head self-attention mechanism that allows the model to capture different aspects of the data. The attention output is combined with the original embeddings using a residual connection followed by layer normalization.
2. A feed-forward neural network (FFN) with two dense layers is applied after each attention mechanism, also followed by a residual connection and layer normalization.
<p>
<strong>Outputs:</strong> After processing the input through the specified number of encoder layers, the function applies a dense layer with softmax activation to produce a probability distribution over the vocabulary for each position in the input sequence.
<p>
<strong>Model Creation:</strong> Finally, it encapsulates the complete architecture within a Keras Model object, which can be compiled and trained.

</div>

In [9]:
def BERT(vocab_size=26, embed_size=128, n_layers=2, n_heads=4, max_length=256):
    input_seq = tf.keras.layers.Input(shape=(max_length,), dtype='int32')

    embedding_layer = tf.keras.layers.Embedding(input_dim=vocab_size, output_dim=embed_size, input_length=max_length)
    embedded_seq = embedding_layer(input_seq)

    x = embedded_seq
    for _ in range(n_layers):
        attention_output = MultiSelfAttentionHead(num_heads=n_heads, key_dim=embed_size // n_heads)(x)
        x = tf.keras.layers.Add()([attention_output, x])  # Residual connection
        x = tf.keras.layers.LayerNormalization(epsilon=1e-6)(x)
        
        ffn_output = tf.keras.Sequential([
            tf.keras.layers.Dense(embed_size, activation='relu'),
            tf.keras.layers.Dense(embed_size)
        ])(x)
        x = tf.keras.layers.Add()([ffn_output, x])  # Residual connection
        x = tf.keras.layers.LayerNormalization(epsilon=1e-6)(x)

    outputs = tf.keras.layers.Dense(vocab_size, activation='softmax')(x)

    model = tf.keras.Model(inputs=input_seq, outputs=outputs)
    return model

# Create the BERT model with the specified parameters
bert_model = BERT()
bert_model.summary()


Model: "model"
__________________________________________________________________________________________________
 Layer (type)                Output Shape                 Param #   Connected to                  
 input_1 (InputLayer)        [(None, 256)]                0         []                            
                                                                                                  
 embedding (Embedding)       (None, 256, 128)             3328      ['input_1[0][0]']             
                                                                                                  
 multi_self_attention_head   (None, None, 128)            16384     ['embedding[0][0]']           
 (MultiSelfAttentionHead)                                                                         
                                                                                                  
 add (Add)                   (None, 256, 128)             0         ['multi_self_attention_he

<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">
<h3 style='color:#2c3e50;'><strong>Section 3: Implementing Noised Tokens Prediction (NTP)</strong></h3>

<p style='color:#2c3e50;'>In previous cell, you implemented BERT architecture. In main version of BERT, we use Masked Language Models (MLMs) and Next Sentence Prediction (NSP) as pretraining tasks. 
<p>
In Masked Language Modeling, random tokens in a sentence are replaced with a special token (e.g., "[MASK]"). The model's task is to predict the original token based on the context provided by the other non-masked tokens in the input. This allows the model to develop a deep understanding of language structure and word relationships. 
<p>
The Next Sentence Prediction task involves presenting the model with two sentences and requiring it to predict whether the second sentence follows the first in the original text. This helps the model to learn relationships between consecutive sentences, which is important for understanding the coherence and flow of paragraphs. 
<br>
You can read more about BERT pretraining tasks in the original <a href="https://arxiv.org/pdf/1810.04805.pdf">paper</a> or <a href="https://towardsdatascience.com/bert-explained-state-of-the-art-language-model-for-nlp-f8b21a9b6270">here</a>.
<p>
In this assignment, however, we want to define a new task called <b>Noised Tokens Prediction (NTP)</b> and use it instead of MLMs and NSP tasks. In this task, we randomlely change tokens in the input sequence to get a noisy sequence to get it to the model and want the model to predict the original sequence. For example, if the input sequence is "ABCDEF", with some probability, it is changed into "ABXDEF"; so the input of the model would be "ABXDEF" and the output should be "ABCDEF". You need to implement this task in the following cell. Feel free to use numpy library for randomization.</p>
</p>

</div>

In [10]:

def noise_tokens(encoded_seqs, p_seq_noise=0.15, n_tokens=26, max_length=256):
    '''
    Add noise to the input sequences.
    Args:
        encoded_seqs (np.array): encoded sequences with shape [Batch_size, max_length]
        p_seq_noise (float): probability of adding noise to each token
        n_tokens (int): number of tokens in the vocabulary
        max_length (int): maximum length of the sequence
    Returns:
        noisy_encoded_seqs (np.array): encoded sequences with noise with shape [Batch_size, max_length]
        encoded_seqs (np.array): original encoded sequences with shape [Batch_size, max_length]
    '''
    noisy_encoded_seqs = encoded_seqs.copy()  # Make a copy to add noise
    # Generate random positions for applying noise
    noise_mask = np.random.rand(*encoded_seqs.shape) < p_seq_noise
    # Generate random tokens to insert for noise
    random_tokens = np.random.randint(1, n_tokens, size=encoded_seqs.shape)
    # Apply noise where the noise mask is True
    noisy_encoded_seqs[noise_mask] = random_tokens[noise_mask]
    
    # Ensure the sequences are of max_length
    if noisy_encoded_seqs.shape[1] > max_length:
        noisy_encoded_seqs = noisy_encoded_seqs[:, :max_length]
    elif noisy_encoded_seqs.shape[1] < max_length:
        padding = np.zeros((noisy_encoded_seqs.shape[0], max_length - noisy_encoded_seqs.shape[1]), dtype=np.int32)
        noisy_encoded_seqs = np.concatenate([noisy_encoded_seqs, padding], axis=1)
    
    return noisy_encoded_seqs, encoded_seqs


<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h3 style='color:#2c3e50;'><strong>Section 4: Pretraining the Model</strong></h3>

<p style='color:#2c3e50;'>Create an instance of your BERT model and train it on pretrain data.</p>

</div>

In [11]:
def pretraining(model, noisy_pretrain_seq, pretrain_seq, epochs, batch_size, lr=1e-2, max_length=256):
    '''
    Pretrain the model: Input is noised sequences and output is original sequences.
    Args:
        model (tf.keras.Model): BERT model
        noisy_pretrain_seq (np.array): encoded sequences with noise with shape [Batch_size, max_length]
        pretrain_seq (np.array): original encoded sequences with shape [Batch_size, max_length]  
        epochs (int): number of epochs
        batch_size (int): batch size 
        lr (float): learning rate
        max_length (int): maximum length of sequences, used for truncating/padding
    Returns:
        model (tf.keras.Model): pretrained BERT model
    '''

    # Truncate sequences if they are longer than max_length
    noisy_pretrain_seq = noisy_pretrain_seq[:, :max_length]
    pretrain_seq = pretrain_seq[:, :max_length]

    # Print shapes for debugging
    print("Shape of noisy_pretrain_seq:", noisy_pretrain_seq.shape)
    print("Shape of pretrain_seq:", pretrain_seq.shape)

    # Loss function and optimizer
    loss_function = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
    optimizer = tf.keras.optimizers.Adam(learning_rate=lr)

    # Compile the model
    model.compile(optimizer=optimizer, loss=loss_function, metrics=['accuracy'])

    # Train the model
    model.fit(
        noisy_pretrain_seq,
        pretrain_seq,
        epochs=epochs,
        batch_size=batch_size
    )

    return model



In [12]:
# please don't change this cell and don't add any print statements to this cell. Run it without changes.
# this cell is for training the model on the pretrain data.

model = BERT(vocab_size=vocab_size, embed_size=128, n_layers=2, n_heads=4, max_length=max_length)
model.summary()

noisy_pretrain_seq, pretrain_seq = noise_tokens(pretrain_seq)
pretrained_model = pretrained_model = pretraining(model, noisy_pretrain_seq, pretrain_seq, epochs, batch_size, lr, max_length)

Model: "model_1"
__________________________________________________________________________________________________
 Layer (type)                Output Shape                 Param #   Connected to                  
 input_2 (InputLayer)        [(None, 256)]                0         []                            
                                                                                                  
 embedding_1 (Embedding)     (None, 256, 128)             3328      ['input_2[0][0]']             
                                                                                                  
 multi_self_attention_head_  (None, None, 128)            16384     ['embedding_1[0][0]']         
 2 (MultiSelfAttentionHead)                                                                       
                                                                                                  
 add_4 (Add)                 (None, 256, 128)             0         ['multi_self_attention_h

C:\Users\mahee\anaconda3\envs\Jupyter\lib\site-packages\keras\src\backend.py:5727: UserWarning: "`sparse_categorical_crossentropy` received `from_logits=True`, but the `output` argument was produced by a Softmax activation and thus does not represent logits. Was this intended?
  output, from_logits = _get_logits(



938/938 [==============================] - 135s 138ms/step - loss: 0.7338 - accuracy: 0.8575
Epoch 2/10
938/938 [==============================] - 131s 139ms/step - loss: 0.7248 - accuracy: 0.8587
Epoch 3/10
938/938 [==============================] - 129s 137ms/step - loss: 1.4986 - accuracy: 0.5718
Epoch 4/10
938/938 [==============================] - 128s 137ms/step - loss: 2.4896 - accuracy: 0.2151
Epoch 5/10
938/938 [==============================] - 131s 140ms/step - loss: 2.2626 - accuracy: 0.2445
Epoch 6/10
938/938 [==============================] - 131s 140ms/step - loss: 2.2512 - accuracy: 0.2377
Epoch 7/10
938/938 [==============================] - 130s 138ms/step - loss: 2.0798 - accuracy: 0.3096
Epoch 8/10
938/938 [==============================] - 130s 138ms/step - loss: 2.0173 - accuracy: 0.3707
Epoch 9/10
938/938 [==============================] - 129s 137ms/step - loss: 2.1703 - accuracy: 0.2886
Epoch 10/10
938/938 [==============================] - 130s 139ms/step - l

In [13]:
# please don't change this cell and don't add any print statements to this cell. Run it without changes.
# this cell is for saving the pretrained model.

if not os.path.exists('model'):
    os.makedirs('model')    
pretrained_model.save_weights('model/model_pretrained.h5')

<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h2 style='color:#2c3e50;'><strong>Section 5: Fine-tuning Large Language Models</strong></h2>

<hr style="border-top: 2px solid #3498db; margin-bottom: 20px; margin-top: 20px;">

<ul style='color:#2c3e50;'>
    <li>Prepare the data for fine-tuning: Tokenize protein sequences in both train and test data of signal peptides detection and structure classification tasks. Then, convert them to numpy arrays. (like what you did in section 1 for pretraining data)</li>
    <li>Fine-tune the model on tasks signal peptides detection (binary classification) and structure classification (multi-class classification). By fine-tuning, we mean the process of adjusting the parameters of the pre-trained model to suit a specific task or dataset. Through this process, we can enhance the model's ability to perform a specialized task more accurately.</li>
    <li>For each task, you need to call the pretrained model and add a classification layer on top of it. Then, you need to freeze the pretrained model and train the classification layer for 10 epochs. After that, you need to unfreeze the pretrained model and train the whole model for 5 epochs.</li>
    </li>
</ul>

</div>

In [14]:
# please don't change this cell and don't add any print statements to this cell. Run it without changes.
# this cell is for connecting the layers of the pretrained model to the layers of the finetuned model. 
# the output of the last layer of the pretrained model is considered as the input of the finetuned model. 

def concat_layers_for_finetuning(model):
    # Create a list to hold the layers to be concatenated
    return tf.keras.Model(inputs=model.inputs, outputs=model.layers[-1].output)

In [15]:
# please don't change this cell
# this cell is for loading the pretrained model and keep the output of the last layer to be used for finetuning.

model_SP = BERT(vocab_size=vocab_size, embed_size=embed_size, n_layers=n_layers, n_heads=n_heads, max_length=max_length)
model_SP.load_weights('model/model_pretrained.h5')
concat_model_SP = concat_layers_for_finetuning(model_SP)

In [16]:
def finetune_peptide_detection(concat_model_SP, signalp_train_data, epochs=10, batch_size=32, lr=1e-2, n_classes=7):
    # 1. Tokenize the sequences in signalp_train_data
    tokenized_seqs = np.array([tokenize_seq(seq, max_length=max_length) for seq in signalp_train_data['seq']])
    tokenized_seqs = tokenized_seqs.astype(np.int32)

    # 2. Make the labels in signalp_train_data as a numpy array
    labels = signalp_train_data['label'].to_numpy()

    # 3. Make the layers of the concat_model_SP non-trainable
    for layer in concat_model_SP.layers:
        layer.trainable = False

     # 4. Add a pooling layer and a dense layer with a proper activation to the end of the model_concat_SC
    pooled_output = tf.keras.layers.GlobalAveragePooling1D()(concat_model_SP.output)
    output_layer = tf.keras.layers.Dense(n_classes, activation='softmax')(pooled_output)

    # 5. Define a new model with the input of concat_model_SP and the output of the dense layer
    fine_tuned_model = tf.keras.Model(inputs=concat_model_SP.input, outputs=output_layer)

    # 6. Define the loss function and optimizer
    loss_function = tf.keras.losses.SparseCategoricalCrossentropy()
    optimizer = tf.keras.optimizers.Adam(learning_rate=lr)

    # 7. Compile the model with the loss function and optimizer
    fine_tuned_model.compile(optimizer=optimizer, loss=loss_function, metrics=['accuracy'])

    # 8. Fit the model on the training data for 10 epochs (initial phase of fine-tuning)
    fine_tuned_model.fit(tokenized_seqs, labels, epochs=epochs, batch_size=batch_size, validation_split=0.1)

    # 9. Make all layers of the model trainable
    
    for layer in fine_tuned_model.layers:
        layer.trainable = True

    # 10. Fit the model on the training data for additional epochs (full fine-tuning)
    fine_tuned_model.fit(tokenized_seqs, labels, epochs=5, batch_size=batch_size, validation_split=0.1)

    # 11. Save the new model in the model directory
    if not os.path.exists('model'):
        os.makedirs('model')
    fine_tuned_model.save('model/model_finetuned_SP.h5')


finetune_peptide_detection(concat_model_SP, signalp_train_data, epochs=10, batch_size=32, lr=1e-2, n_classes=7)



Epoch 1/10
468/468 [==============================] - 33s 63ms/step - loss: 0.6052 - accuracy: 0.8360 - val_loss: 0.4621 - val_accuracy: 0.8320
Epoch 2/10
468/468 [==============================] - 27s 58ms/step - loss: 0.4449 - accuracy: 0.8381 - val_loss: 0.4500 - val_accuracy: 0.8320
Epoch 3/10
468/468 [==============================] - 26s 56ms/step - loss: 0.4386 - accuracy: 0.8381 - val_loss: 0.4468 - val_accuracy: 0.8320
Epoch 4/10
468/468 [==============================] - 28s 60ms/step - loss: 0.4363 - accuracy: 0.8381 - val_loss: 0.4447 - val_accuracy: 0.8320
Epoch 5/10
468/468 [==============================] - 27s 58ms/step - loss: 0.4348 - accuracy: 0.8381 - val_loss: 0.4436 - val_accuracy: 0.8320
Epoch 6/10
468/468 [==============================] - 28s 59ms/step - loss: 0.4339 - accuracy: 0.8381 - val_loss: 0.4440 - val_accuracy: 0.8320
Epoch 7/10
468/468 [==============================] - 28s 59ms/step - loss: 0.4334 - accuracy: 0.8381 - val_loss: 0.4425 - val_accuracy:

C:\Users\mahee\anaconda3\envs\Jupyter\lib\site-packages\keras\src\engine\training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


In [17]:
# please don't change this cell
# this cell is for loading the pretrained model and keep the output of the last layer to be used for finetuning.

model_SC = BERT(vocab_size=vocab_size, embed_size=128, n_layers=2, n_heads=4, max_length=max_length)
model_SC.load_weights('model/model_pretrained.h5')
model_concat_SC = concat_layers_for_finetuning(model_SC)

In [18]:
def finetune_structure_classification(model_concat_SC, scop_train_data, epochs=10, batch_size=32, lr=1e-2, n_classes=7):
    # 1. Tokenize the sequences in scop_train_data
    tokenized_seqs = np.array([tokenize_seq(seq, max_length=max_length) for seq in scop_train_data['seq']])
    tokenized_seqs = tokenized_seqs.astype(np.int32)

    # 2. Make the labels in scop_train_data as a numpy array
    labels = pd.get_dummies(scop_train_data['label']).to_numpy()  # One-hot encoding for multi-class classification

    # 3. Make the layers of the model_concat_SC non-trainable
    for layer in model_concat_SC.layers:
        layer.trainable = False

    # 4. Add a pooling layer and a dense layer with a proper activation to the end of the model_concat_SC
    pooled_output = tf.keras.layers.GlobalAveragePooling1D()(model_concat_SC.output)
    output_layer = tf.keras.layers.Dense(n_classes, activation='softmax')(pooled_output)


    # 5. Define a new model with the input of model_concat_SC and the output of the dense layer
    fine_tuned_model = tf.keras.Model(inputs=model_concat_SC.input, outputs=output_layer)

    # 6. Define the loss function and optimizer
    loss_function = tf.keras.losses.CategoricalCrossentropy()
    optimizer = tf.keras.optimizers.Adam(learning_rate=lr)

    # 7. Compile the model with the loss function and optimizer
    fine_tuned_model.compile(optimizer=optimizer, loss=loss_function, metrics=['accuracy'])

    # 8. Fit the model on the training data for 10 epochs (initial phase of fine-tuning)
    fine_tuned_model.fit(tokenized_seqs, labels, epochs=epochs, batch_size=batch_size, validation_split=0.1)

    # 9. Make all layers of the model trainable
    for layer in fine_tuned_model.layers:
        layer.trainable = True

    # 10. Fit the model on the training data for additional epochs (full fine-tuning)
    fine_tuned_model.fit(tokenized_seqs, labels, epochs=5, batch_size=batch_size, validation_split=0.1)

    # 11. Save the new model in the model directory
    if not os.path.exists('model'):
        os.makedirs('model')
    fine_tuned_model.save('model/model_finetuned_SC.h5')


finetune_structure_classification(model_concat_SC, scop_train_data, epochs=10, batch_size=32, lr=1e-2, n_classes=7)



Epoch 1/10
441/441 [==============================] - 29s 61ms/step - loss: 1.6113 - accuracy: 0.3090 - val_loss: 1.5719 - val_accuracy: 0.3227
Epoch 2/10
441/441 [==============================] - 27s 61ms/step - loss: 1.5488 - accuracy: 0.3366 - val_loss: 1.5401 - val_accuracy: 0.3546
Epoch 3/10
441/441 [==============================] - 27s 61ms/step - loss: 1.5253 - accuracy: 0.3667 - val_loss: 1.5209 - val_accuracy: 0.3916
Epoch 4/10
441/441 [==============================] - 28s 63ms/step - loss: 1.5098 - accuracy: 0.3828 - val_loss: 1.5105 - val_accuracy: 0.3731
Epoch 5/10
441/441 [==============================] - 27s 61ms/step - loss: 1.4999 - accuracy: 0.3839 - val_loss: 1.5012 - val_accuracy: 0.4031
Epoch 6/10
441/441 [==============================] - 26s 58ms/step - loss: 1.4919 - accuracy: 0.3906 - val_loss: 1.4925 - val_accuracy: 0.3967
Epoch 7/10
441/441 [==============================] - 25s 57ms/step - loss: 1.4859 - accuracy: 0.3951 - val_loss: 1.4885 - val_accuracy:

C:\Users\mahee\anaconda3\envs\Jupyter\lib\site-packages\keras\src\engine\training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h2 style='color:#2c3e50;'><strong>Section 6: Prediction and Evaluation</strong></h2>

<hr style="border-top: 2px solid #3498db; margin-bottom: 20px; margin-top: 20px;">

<ul style='color:#2c3e50;'>
    <li>After finetuning, do <b>prediction</b> task on test data. Then, report <b>f1 score, precision, recall, and accuracy </b> for both datasets.
    You can read more about these metrics <a href="https://scikit-learn.org/stable/modules/model_evaluation.html#classification-metrics">here</a>.</li>
</ul>

</div>

<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h3 style='color:#2c3e50;'><strong>Task 6.1: Prediction </strong></h3>

<p style='color:#2c3e50;'> Make prediction on both test datasets.</p>

</div>

In [19]:

# Load the finetuned model, specifying the custom object
model_finetuned_SP = tf.keras.models.load_model('model/model_finetuned_SP.h5', custom_objects={'MultiSelfAttentionHead': MultiSelfAttentionHead})

# Tokenize the sequences in the signal peptide test data
tokenized_test_seqs = np.array([tokenize_seq(seq, max_length=max_length) for seq in signalp_test_data['seq']])
tokenized_test_seqs = tokenized_test_seqs.astype(np.int32)

# Predict using the finetuned model
y_pred_SP_probs = model_finetuned_SP.predict(tokenized_test_seqs)
# Convert probabilities to class labels (0 or 1)
y_pred_SP = np.argmax(y_pred_SP_probs, axis=1)

# Save the predicted labels, true labels, and sequences to a CSV file
signalp_test_data['pred_label'] = y_pred_SP
signalp_test_data.to_csv('data/signal_peptide/signalp_pred.csv', index=False)


130/130 [==============================] - 7s 50ms/step


In [20]:
# Load the finetuned model for SCOP classification
model_finetuned_SC = tf.keras.models.load_model('model/model_finetuned_SC.h5', custom_objects={'MultiSelfAttentionHead': MultiSelfAttentionHead})

# Tokenize the sequences in the SCOP test data
# Assuming tokenize_seq function and max_length are defined as provided earlier
tokenized_test_seqs = np.array([tokenize_seq(seq, max_length=max_length) for seq in scop_test_data['seq']])
tokenized_test_seqs = tokenized_test_seqs.astype(np.int32)

# Predict using the finetuned model
y_pred_SC_probs = model_finetuned_SC.predict(tokenized_test_seqs)
# Convert probabilities to class labels
y_pred_SC = np.argmax(y_pred_SC_probs, axis=1)

# Save the predicted labels, true labels, and sequences to a CSV file
scop_test_data['pred_label'] = y_pred_SC
scop_test_data.to_csv('data/scop/scop_pred.csv', index=False)

123/123 [==============================] - 6s 47ms/step


<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h3 style='color:#2c3e50;'><strong>Task 6.2: Evaluation</strong></h3>

<p style='color:#2c3e50;'>Calculate F1 score, precision, recall and accuracy for both test datasets.</p>

</div>

In [21]:
# Load the CSV file into a Pandas DataFrame
signalp_test_data = pd.read_csv('data/signal_peptide/signalp_pred.csv')

# Retrieve the top 10 rows
top_10_signalp_test_data = signalp_test_data.head(10)

# Display the top 10 rows
print(top_10_signalp_test_data)

   label                                                seq  pred_label
0      0  MAVSVTPIRDTKWLTLEVCREFQRGTCSRPDTECKFAHPSKSCQVE...           0
1      0  MGSSFFGRPKMGGSSSSSPTSSSSSPAKRGKNKNGSDKPKQPQRGL...           0
2      0  MDISTRSKDPGSAERTAQKRKFPSPPHSSNGHSPQDTSTSPIKKKK...           0
3      0  MTATIIAAITITTIVIITPMEEITTMTIPMEEITTMTIPMEEITTM...           0
4      0  MAFKFKTFAAVGALIGSLALAGCGQDEKDPNHIKVGVIVGAEQQVA...           0
5      0  MDQNKSTQATGNGKKTEEINELLESFIKEGPKLLWGSTNLKYEDQI...           0
6      0  MFGTPSRRTFLTASALSAMALAASPTVTDAIAAPGPDSWSALCERW...           0
7      0  MFSKYLQSRVCGLHSFTNSSAQQLFSKSIAHSSRRNFVISSSCTKF...           0
8      0  MEEKPGQPQPQHHHSHHHPHHHPQQQQQQQSHHHHHYYFYNHSHNH...           0
9      0  MSASSRFIPEHRRQNYKGKGTFQADELRRRRETQQIEIRKQKREEN...           0


In [22]:
# Load the CSV file into a Pandas DataFrame
scop_pred_data = pd.read_csv('data/scop/scop_pred.csv')

# Retrieve the top 10 rows
top_10_scop_pred = scop_pred_data.head(10)

# Display the top 10 rows
print(top_10_scop_pred)

                                                 seq label  pred_label
0           DPMTCEQAMASCEHTMCGYCQGPLYMTCIGITTDPECGLP     a           2
1  GSDKIHHHHHHMNIFEAIENRHSVRDFLERKMPERVKDDIENLLVK...     d           2
2  MPDLNSSTDSAASASAASDVSVESTAEATVCTVTLEKMSAGLGFSL...     b           3
3  MDFHIRKATNSDAEAIQHVATTSWHHTYQDLIPSDVQDDFLKRFYN...     d           2
4  DQSGYERGLTLPLRHPSGLFDGETEAVWGLNTAYSVVEKSVSTRDY...     b           2
5  ADVILFPSKNGAVTFTHKRHSEFVRECRSCHEKTPGKIRNFGKDYA...     a           3
6  MTPKLKLNNNINWTKRTIDSLFDLKKGEMLEKELITPEGKYEYFNG...     d           1
7  TEGLSDKEQRFVDKLYTGLIQGQRACLAEAITLVESTHSRKKELAQ...     c           2
8  MGMIVFVRFNSSYGFPVEVDSDTSILQLKEVVAKRQGVPADQLRVI...     d           3
9  PQQCDQTFTIATTDYAMQTILPFALPRIYQEAPNVSFNFLPLQHDR...     c           3


In [23]:
# Assuming signalp_test_data is your DataFrame containing the true labels and predicted labels
y_true_SP = signalp_test_data['label'].values
y_pred_SP = signalp_test_data['pred_label'].values

accuracy_SP = accuracy_score(y_true_SP, y_pred_SP)
precision_SP = precision_score(y_true_SP, y_pred_SP)
recall_SP = recall_score(y_true_SP, y_pred_SP)
f1_score_SP = f1_score(y_true_SP, y_pred_SP)

print("Accuracy:", accuracy_SP)
print("Precision:", precision_SP)
print("Recall:", recall_SP)
print("F1 Score:", f1_score_SP)

Accuracy: 0.8376685934489403
Precision: 0.0
Recall: 0.0
F1 Score: 0.0


C:\Users\mahee\anaconda3\envs\Jupyter\lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [24]:
from sklearn.preprocessing import LabelEncoder

# Create a label encoder object
label_encoder = LabelEncoder()

# Fit the label encoder on the true labels and transform them to numeric
y_true_SC_encoded = label_encoder.fit_transform(scop_pred_data['label'].values)

# Now, y_true_SC_encoded and y_pred_SC are both numeric and can be used to calculate metrics
accuracy_SC = accuracy_score(y_true_SC_encoded, y_pred_SC)
precision_SC = precision_score(y_true_SC_encoded, y_pred_SC, average='micro')
recall_SC = recall_score(y_true_SC_encoded, y_pred_SC, average='micro')
f1_score_SC = f1_score(y_true_SC_encoded, y_pred_SC, average='micro')

print("Accuracy:", accuracy_SC)
print("Precision:", precision_SC)
print("Recall:", recall_SC)
print("F1 Score:", f1_score_SC)


Accuracy: 0.39836776332568224
Precision: 0.39836776332568224
Recall: 0.39836776332568224
F1 Score: 0.39836776332568224


<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h3 style='color:#2c3e50;'><strong>Section 7: Experimentation</strong></h3>

<p style='color:#2c3e50;'>In the realm of natural language processing, one essential aspect is the reliance on vast datasets for training language models. With limited access to resources like GPUs or memory, we cannot pretrain our own model in this assignment on a huge dataset. Instead, we can use available pretrained models for protein sequences to check if a LLM can improve the performance. In this section of our study, we aim to utilize a pre-trained language model named <b>DistilProtBert</b>, which is accessible through the popular <b>Transformers</b> library. 
The key objective here is to employ this pre-trained DistilProtBert model and fine-tune it, just as we did in the previous section and compare the performance of the fine-tuned model with and without using pretrained model.
Once you have completed the fine-tuning process, it is important to analyze and record any noticeable differences in the model's performance. This entails a comprehensive evaluation of how the fine-tuned model compares to its performance before fine-tuning. You should pay attention to various metrics such as accuracy, precision, recall, and F1-score, among others, to understand the overall impact of the fine-tuning process.
In this section, you are free to use any libraries or pretrained models. However, the smallest pretrained model available for preotein sequences is DistilProtBert. 
</p>
<p>
You can read more about DistilProtBert <a href="https://huggingface.co/yarongef/DistilProtBert">here</a>.
<p> What you need to do in this section is: </p>
<p>
1. Select one of two finetuning tasks from data directory.</p>
<p>
2. Finetune DistilProtBert on the selected dataset and report the performance of the model on test data. You need to report <b>f1 score, precision, recall, and accuracy </b>. </p>
<p>
3. Train your classifier of previous step without using pretrained model (DisilProtBert). Report the performance of the model on test data. You need to report <b>f1 score, precision, recall, and accuracy </b>.
</p>
<p>
4. Compare the performance of the two models and report your findings.
</p>
</div>

In [25]:
%pip install transformers[torch]
%pip install accelerate -U

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


In [26]:
from transformers import DistilBertModel, DistilBertTokenizer, \
Trainer, TrainingArguments, DistilBertConfig
import torch
from torch.utils.data import Dataset, DataLoader
import torch.nn.functional as F
import pandas as pd
import numpy as np
import torch.nn as nn
from sklearn.preprocessing import LabelEncoder

# Load the pretrained model and tokenizer from yarongef/DistilProtBert
tokenizer = DistilBertTokenizer.from_pretrained("yarongef/DistilProtBert", do_lower_case=False)
model_distilprotbert = DistilBertModel.from_pretrained("yarongef/DistilProtBert")

The tokenizer class you load from this checkpoint is not the same type as the class this function is called from. It may result in unexpected tokenization. 
The tokenizer class you load from this checkpoint is 'BertTokenizer'. 
The class this function is called from is 'DistilBertTokenizer'.
You are using a model of type bert to instantiate a model of type distilbert. This is not supported for all configurations of models and can yield errors.
Some weights of DistilBertModel were not initialized from the model checkpoint at yarongef/DistilProtBert and are newly initialized: ['transformer.layer.12.ffn.lin2.weight', 'transformer.layer.9.attention.v_lin.weight', 'transformer.layer.10.attention.out_lin.bias', 'transformer.layer.14.ffn.lin2.weight', 'transformer.layer.4.ffn.lin2.weight', 'transformer.layer.4.attention.k_lin.bias', 'transformer.layer.8.ffn.lin2.weight', 'transformer.layer.14.attention.out_lin.weight', 'transformer.layer.1.ffn.lin1.bias', 'transformer.layer.12.ffn.lin1.bias',

In [27]:
# Load the datasets
scop_train_data = pd.read_csv('data/scop/scop.train.csv')
scop_test_data = pd.read_csv('data/scop/scop.test.csv')


In [28]:
print(scop_train_data.columns)
print(scop_test_data.columns)


Index(['seq', 'label'], dtype='object')
Index(['seq', 'label'], dtype='object')


In [29]:
# Tokenize the train data
train_encodings = tokenizer(scop_train_data['seq'].tolist(), truncation=True, padding=True, max_length=512)

# Tokenize the test data
test_encodings = tokenizer(scop_test_data['seq'].tolist(), truncation=True, padding=True, max_length=512)


In [30]:
label_encoder = LabelEncoder()
scop_train_data['label'] = label_encoder.fit_transform(scop_train_data['label'])
scop_test_data['label'] = label_encoder.transform(scop_test_data['label'])


In [31]:
class ScopDataset(Dataset):
    def __init__(self, encodings, labels):
        self.encodings = encodings
        self.labels = labels

    def __getitem__(self, idx):
        item = {key: torch.tensor(val[idx]) for key, val in self.encodings.items()}
        item['labels'] = torch.tensor(self.labels[idx])
        return item

    def __len__(self):
        return len(self.labels)

# Convert labels to integers if they're not already
scop_train_data['label'] = scop_train_data['label'].astype(int)
scop_test_data['label'] = scop_test_data['label'].astype(int)

# Create dataset objects
train_dataset = ScopDataset(train_encodings, scop_train_data['label'].tolist())
test_dataset = ScopDataset(test_encodings, scop_test_data['label'].tolist())


In [32]:
from transformers import DistilBertForSequenceClassification

num_labels = len(scop_train_data['label'].unique())
model = DistilBertForSequenceClassification.from_pretrained("yarongef/DistilProtBert", num_labels=num_labels)


You are using a model of type bert to instantiate a model of type distilbert. This is not supported for all configurations of models and can yield errors.
Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at yarongef/DistilProtBert and are newly initialized: ['transformer.layer.12.ffn.lin2.weight', 'transformer.layer.9.attention.v_lin.weight', 'transformer.layer.10.attention.out_lin.bias', 'transformer.layer.14.ffn.lin2.weight', 'transformer.layer.4.ffn.lin2.weight', 'transformer.layer.4.attention.k_lin.bias', 'transformer.layer.8.ffn.lin2.weight', 'transformer.layer.14.attention.out_lin.weight', 'transformer.layer.1.ffn.lin1.bias', 'transformer.layer.12.ffn.lin1.bias', 'embeddings.LayerNorm.weight', 'transformer.layer.5.ffn.lin1.bias', 'transformer.layer.11.sa_layer_norm.bias', 'transformer.layer.4.output_layer_norm.bias', 'transformer.layer.5.ffn.lin2.bias', 'transformer.layer.14.sa_layer_norm.weight', 'transformer.layer.13.attention.k

In [33]:
from transformers import Trainer, TrainingArguments

training_args = TrainingArguments(
    output_dir='./results',          
    num_train_epochs=3,          
    per_device_train_batch_size=16,
    warmup_steps=500,
    weight_decay=0.01,
    logging_dir='./logs',
    logging_steps=10,
)


trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=test_dataset
)

trainer.train()


TrainOutput(global_step=2940, training_loss=1.6393194626788703, metrics={'train_runtime': 5801.0662, 'train_samples_per_second': 8.109, 'train_steps_per_second': 0.507, 'total_flos': 134230134648960.0, 'train_loss': 1.6393194626788703, 'epoch': 3.0})

In [35]:
from sklearn.metrics import accuracy_score
# Evaluate the model
evaluation_results = trainer.evaluate(test_dataset)

# 'evaluation_results' will contain metrics like loss and accuracy
print("Evaluation Results:", evaluation_results)

# To calculate precision, recall, and F1 score, we need to get predictions
predictions = trainer.predict(test_dataset)

# Extract the predicted labels and true labels
predicted_labels = np.argmax(predictions.predictions, axis=1)
true_labels = predictions.label_ids


precision = precision_score(true_labels, predicted_labels, average='weighted', zero_division=1)

recall = recall_score(true_labels, predicted_labels, average='weighted')
f1 = f1_score(true_labels, predicted_labels, average='weighted')


accuracy = accuracy_score(true_labels, predicted_labels)

# Print the results
print(f"Accuracy: {accuracy}")
print(f"Precision: {precision}")
print(f"Recall: {recall}")
print(f"F1 Score: {f1}")


Evaluation Results: {'eval_loss': 1.623274564743042, 'eval_runtime': 72.4633, 'eval_samples_per_second': 54.11, 'eval_steps_per_second': 6.776, 'epoch': 3.0}
Accuracy: 0.2973731191022698
Precision: 0.791057652862343
Recall: 0.2973731191022698
F1 Score: 0.13632280592618307


In [36]:
def encode_sequence(seq, max_length):
    encoding = {char: idx for idx, char in enumerate(sorted(set(''.join(seq))))}
    encoded = [encoding[char] for char in seq]
    # Pad the sequences
    padded = np.zeros(max_length, dtype=np.int64)
    padded[:len(encoded)] = encoded
    return padded

max_seq_length = max(scop_train_data['seq'].apply(len).max(), scop_test_data['seq'].apply(len).max())

scop_train_data['encoded_seq'] = scop_train_data['seq'].apply(lambda x: encode_sequence(x, max_seq_length))
scop_test_data['encoded_seq'] = scop_test_data['seq'].apply(lambda x: encode_sequence(x, max_seq_length))


In [37]:
class SCOPDataset(Dataset):
    def __init__(self, sequences, labels):
        self.sequences = sequences
        self.labels = labels

    def __len__(self):
        return len(self.sequences)

    def __getitem__(self, idx):
        # Cast sequences to float
        return torch.tensor(self.sequences[idx], dtype=torch.float), torch.tensor(self.labels[idx])



# Create Dataset
train_dataset = SCOPDataset(scop_train_data['encoded_seq'].tolist(), scop_train_data['label'].tolist())
test_dataset = SCOPDataset(scop_test_data['encoded_seq'].tolist(), scop_test_data['label'].tolist())

# Create DataLoader
train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=32)


In [38]:
class SimpleClassifier(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim):
        super(SimpleClassifier, self).__init__()
        self.fc1 = nn.Linear(input_dim, hidden_dim)
        self.relu = nn.ReLU()
        self.fc2 = nn.Linear(hidden_dim, output_dim)
    
    def forward(self, x):
        x = self.fc1(x)
        x = self.relu(x)
        x = self.fc2(x)
        return x

# Assuming a fixed input dimension (e.g., max sequence length)
input_dim = max(len(s) for s in scop_train_data['encoded_seq'])
model = SimpleClassifier(input_dim=max_seq_length, hidden_dim=128, output_dim=len(label_encoder.classes_))



In [39]:
criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

# Training loop
for epoch in range(10):
    for sequences, labels in train_loader:
        optimizer.zero_grad()
        outputs = model(sequences)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
    print(f'Epoch {epoch+1}: Loss = {loss.item()}')


Epoch 1: Loss = 1.0739502906799316
Epoch 2: Loss = 1.2770074605941772
Epoch 3: Loss = 1.2039408683776855
Epoch 4: Loss = 1.27092444896698
Epoch 5: Loss = 1.2900056838989258
Epoch 6: Loss = 1.3462891578674316
Epoch 7: Loss = 1.1318986415863037
Epoch 8: Loss = 1.129727840423584
Epoch 9: Loss = 0.9291144013404846
Epoch 10: Loss = 1.1055880784988403


In [40]:
all_predictions = []
all_labels = []

model.eval()  # Set the model to evaluation mode
with torch.no_grad():
    for sequences, labels in test_loader:
        outputs = model(sequences)
        _, predicted = torch.max(outputs.data, 1)
        
        # Collect predictions and labels
        all_predictions.extend(predicted.numpy())
        all_labels.extend(labels.numpy())

# Calculate metrics
accuracy = accuracy_score(all_labels, all_predictions)
precision = precision_score(all_labels, all_predictions, average='weighted')
recall = recall_score(all_labels, all_predictions, average='weighted')
f1 = f1_score(all_labels, all_predictions, average='weighted')

# Print results
print(f"Accuracy: {accuracy}")
print(f"Precision: {precision}")
print(f"Recall: {recall}")
print(f"F1 Score: {f1}")

Accuracy: 0.42514664626370824
Precision: 0.4026632868863724
Recall: 0.42514664626370824
F1 Score: 0.40655825237241805


Model 2: Finetuned DistilProtBert

Accuracy: 0.2974
Precision: 0.7911
Recall: 0.2974
F1 Score: 0.1363

Model 3: Custom-Trained Classifier

Accuracy: 0.4251
Precision: 0.4027
Recall: 0.4251
F1 Score: 0.4066

Analysis and Findings:

Accuracy: Model 3 outperforms Model 2 in accuracy (0.4251 vs. 0.2974). This suggests that Model 3 is more effective in correctly classifying the test samples overall.

Precision: Model 2 has a higher precision (0.7911) than Model 3 (0.4027). This indicates that when Model 2 predicts a class, it is more likely to be correct compared to Model 3. However, the high precision of Model 2 might be misleading due to its lower recall.

Recall: Both models have similar recall scores, with Model 3 slightly outperforming Model 2 (0.4251 vs. 0.2974). This suggests that Model 3 is slightly better at identifying all relevant instances in the test data.

F1 Score: Model 3 has a significantly higher F1 score (0.4066) compared to Model 2 (0.1363). The F1 score is a harmonic mean of precision and recall, providing a balance between the two. A higher F1 score for Model 3 indicates it has a better balance of precision and recall.

Conclusions:

Model 3 (Custom-Trained Classifier) appears to be the better model for this specific task, based on the overall balance of accuracy, precision, recall, and F1 score.
Model 2 (DistilProtBert), despite its higher precision, seems to struggle with this particular classification task, as indicated by its lower accuracy and F1 score.

<h2>Submission</h2>

<hr style="border-top: 5px solid orange; margin-top: 1px; margin-bottom: 1px"></hr>

<p style="text-align: justify;">You need to submit a Jupyter Notebook (*.ipynb) file that contains your completed code and results.
<p>
Your answers to sections 1-6 will be evaluated based on the correctness of your implementation. Your answer to section 7 will be evaluated based on the correctness of your implementation, the results you report, and the quality of your analysis.

<span>The file name should be in <strong>FirstName_LastName</strong> format</span>.</p>
<p style="text-align: justify;"><span>DO NOT INCLUDE EXTRA FILES, SUCH AS THE INPUT DATASETS</span>, in your submission;</p>
<p style="text-align: justify;">Please download your assignment after submission and make sure it is not corrupted or empty! We will not be responsible for corrupted submissions and will not take a resubmission after the deadline.</p>

Need Help?
If you need help with this assignment, please get in touch with TAs via their emails, or go to their office hours.
You are highly encouraged to ask your question on the designated channel for Assignment o on Microsoft Teams (not necessarily monitored by the instructor/TAs). Feel free to help other students with general questions. However, DO NOT share your solution.<hr style="border-top: 5px solid orange; margin-top: 1px; margin-bottom: 1px"></hr>